### PR TITLE
filter: aws lambda - add async invocation mode

### PR DESCRIPTION
**Pull request by marcomagdy**
Asynchronous Lambda Invocation Support

Signed-off-by: Marco Magdy <mmagdy@gmail.com>

### CODE REVIEW

File: api/envoy/config/filter/http/aws_lambda/v2alpha/aws_lambda.proto [hunk 1]
```diff
@@ -20,13 +20,27 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
 // AWS Lambda filter config
 message Config {
+  enum InvocationMode {
```
Comment (abaptiste) [inline api/envoy/config/filter/http/aws_lambda/v2alpha/aws_lambda.proto:23]: Let's make this `oneof` and add the validation that it's required.

File: source/extensions/filters/http/aws_lambda/config.cc [hunk 2]
```diff
@@ -37,7 +37,21 @@ Http::FilterFactoryCb AwsLambdaFilterFactory::createFilterFactoryFromProtoTyped(
   auto signer = std::make_shared<Extensions::Common::Aws::SignerImpl>(
       service_name, region, std::move(credentials_provider), context.dispatcher().timeSource());
 
-  FilterSettings filter_settings{proto_config.arn(), proto_config.payload_passthrough()};
+  InvocationMode invocation_mode;
+  switch (proto_config.invocation_mode()) {
+    using namespace envoy::extensions::filters::http::aws_lambda::v3;
```
Comment (abaptiste) [inline source/extensions/filters/http/aws_lambda/config.cc]: nit: This looks awkward here

File: api/envoy/config/filter/http/aws_lambda/v2alpha/aws_lambda.proto (hunk 1 above)
Comment (marcomagdy) [reply api/envoy/config/filter/http/aws_lambda/v2alpha/aws_lambda.proto:23]: why?

Comment (marcomagdy) [reply api/envoy/config/filter/http/aws_lambda/v2alpha/aws_lambda.proto:23]: oneof is a union. I don't need a union here. Is that an Envoy recommendation?

Comment (lavignes) [reply api/envoy/config/filter/http/aws_lambda/v2alpha/aws_lambda.proto:23]: One big benefit of the union is that if we ever need config that is specific to a given mode, then we'd put it as fields in those respective union variants.

A pattern I see is:

```
message SomeThingConfig {
    SomethingCommonConfig  common = 0;
    oneof kind {
        SomethingAConfig a = 1;
        SomethingBConfig b = 2;
    }
}
```

File: test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc [hunk 3]
```diff
@@ -417,7 +461,7 @@ TEST_F(AwsLambdaFilterTest, EncodeDataJsonModeStopIterationAndBuffer) {
  * headers while ignoring any HTTP/2 pseudo-headers.
  */
 TEST_F(AwsLambdaFilterTest, EncodeDataJsonModeTransformToHttp) {
-  setupFilter({Arn, false /*passthrough*/});
+  setupFilter({Arn, InvocationMode::Synchronous, false /*passthrough*/});
   filter_->resolveSettings();
   Http::TestResponseHeaderMapImpl headers;
   headers.setStatus(200);
```
Comment (nealp9084) [inline test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc:474]: This seems to be repeated. Is it possible to reduce this code duplication?

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc [hunk 4]
```diff
@@ -35,10 +35,15 @@ namespace {
...
-void setLambdaHeaders(Http::RequestHeaderMap& headers, absl::string_view function_name) {
+void setLambdaHeaders(Http::RequestHeaderMap& headers, absl::string_view function_name,
+                      InvocationMode mode) {
   headers.setMethod(Http::Headers::get().MethodValues.Post);
   headers.setPath(fmt::format("/2015-03-31/functions/{}/invocations", function_name));
-  headers.setCopy(Http::LowerCaseString{"x-amz-invocation-type"}, "RequestResponse");
+  if (mode == InvocationMode::Synchronous) {
+    headers.setCopy(Http::LowerCaseString{"x-amz-invocation-type"}, "RequestResponse");
+  } else {
+    headers.setCopy(Http::LowerCaseString{"x-amz-invocation-type"}, "Event");
```
Comment (nealp9084) [inline source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc]: nit: DRY

File: test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc (hunk 3 above)
Comment (marcomagdy) [reply test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc:474]: The only way I could think of is using parameterized tests.. :/

File: source/extensions/filters/http/aws_lambda/config.cc (hunk 2 above)
Comment (marcomagdy) [reply source/extensions/filters/http/aws_lambda/config.cc]: Agreed. Fixed.

**Pull request closed without merging**